homeassistant/components/synology_dsm/__init__.py

```diff
@@ -146,30 +180,213 @@ async def async_setup(self):
             device_token=self._entry.data.get("device_token"),
         )
 
+        await self._should_fetch_api()
+
         await self._hass.async_add_executor_job(self._fetch_device_configuration)
-        await self.update()
+        await self.async_update()
 
         self._unsub_dispatcher = async_track_time_interval(
             self._hass,
-            self.update,
+            self.async_update,
             timedelta(
                 minutes=self._entry.options.get(
                     CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
                 )
             ),
         )
 
+    async def _should_fetch_api(self):
+
+        entity_reg = await self._hass.helpers.entity_registry.async_get_registry()
+        entity_entries = entity_registry.async_entries_for_config_entry(
+            entity_reg, self._entry.entry_id
+        )
+        # Entities not added yet
+        if not entity_entries:
+            self._with_security = True
+            self._with_storage = True
+            self._with_utilisation = True
+            return
+
+        # Determine which if at least one entity uses specific API
+        self._with_security = False
+        self._with_storage = False
+        self._with_utilisation = False
+        for entity_entry in entity_entries:
+            # Pass disabled entries
+            if entity_entry.disabled_by:
+                continue
+
+            # Check if we should fetch specific APIs
+            if SynoCoreSecurity.API_KEY in entity_entry.unique_id:
+                self._with_security = True
+                continue
+
+            if SynoStorage.API_KEY in entity_entry.unique_id:
+                self._with_storage = True
+                continue
+
+            if SynoCoreUtilization.API_KEY in entity_entry.unique_id:
+                self._with_utilisation = True
+                continue
+
+        # Reset not used API
+        if not self._with_security:
+            self.dsm._security = None  # pylint: disable=protected-access
```

Comment:
We should change the upstream package to provide a method to reset this.

Reply:
Added in #35640